atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/OracleShrinkExecutor.java

```diff
@@ -0,0 +1,144 @@
+
+import java.sql.SQLException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Stopwatch;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleDdlConfig;
+import com.palantir.atlasdb.keyvalue.dbkvs.OracleTableNameGetter;
+import com.palantir.atlasdb.keyvalue.impl.TableMappingNotFoundException;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.exception.PalantirSqlException;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.nexus.db.sql.SqlConnection;
+
+public class OracleShrinkExecutor {
+    private static final Logger log = LoggerFactory.getLogger(OracleShrinkExecutor.class);
+
+    private final SqlConnectionSupplier connectionPool;
+    private final ExecutorService executorService;
+    private final OracleTableNameGetter oracleTableNameGetter;
+    private OracleDdlConfig oracleDdlConfig;
+
+    private static Future<Boolean> previousShrinkFuture;
+
+    public OracleShrinkExecutor(
+            SqlConnectionSupplier connectionPool,
+            ExecutorService executorService,
+            OracleTableNameGetter oracleTableNameGetter,
+            OracleDdlConfig oracleDdlConfig) {
+        this.connectionPool = connectionPool;
+        this.executorService = executorService;
+        this.oracleTableNameGetter = oracleTableNameGetter;
+        this.oracleDdlConfig = oracleDdlConfig;
+    }
+
+    public void shrinkAsync(TableReference tableRef) {
+        final long secondsToWaitBeforeShrink = getSecondsToWaitBeforeShrink();
+        previousShrinkFuture = executorService.submit(
+                () -> shrinkCompactFollowedByShrink(tableRef, secondsToWaitBeforeShrink));
+
+    }
+
+    private long getSecondsToWaitBeforeShrink() {
+        long secondsToWaitBeforeShrink = oracleDdlConfig.shrinkPauseSeconds();
+        if (previousShrinkFuture != null) {
+            try {
+                if (!previousShrinkFuture.get()) {
+                    secondsToWaitBeforeShrink =
+                            secondsToWaitBeforeShrink + oracleDdlConfig.shrinkPauseOnFailureSeconds();
+                }
+            } catch (InterruptedException | ExecutionException e) {
+                secondsToWaitBeforeShrink = secondsToWaitBeforeShrink + oracleDdlConfig.shrinkPauseOnFailureSeconds();
+            }
+        }
+        return secondsToWaitBeforeShrink;
+    }
+
+    private boolean shrinkCompactFollowedByShrink(TableReference tableRef, long secondsToWaitBeforeShrink) {
+
+        try {
+            TimeUnit.SECONDS.sleep(secondsToWaitBeforeShrink);
+        } catch (InterruptedException e) {
+            log.warn("Skipping Shrink for table: {}", LoggingArgs.tableRef("tableToShrink", tableRef));
```

Comment:
What's the relation between `LoggingArgs` and `(Un)SafeArg`?

It's probably worth saying here that we skipped because the thread was interrupted.

Reply:
Yes, LoggingArgs takes into account the table metadata where you can specify whether a table name is safe to log.